Spark: batch processing (#15839)
dotnet/docs: docs/spark/tutorials/batch-processing.md

---
title: Batch processing with .NET for Apache Spark tutorial
description: Learn how to do batch processing using .NET for Apache Spark.
author: mamccrea
ms.author: mamccrea
ms.date: 12/04/2019
ms.topic: tutorial
---

# Tutorial: Do batch processing with .NET for Apache Spark

In this tutorial, you learn how to do batch processing using .NET for Apache Spark. Batch processing is the transformation of data at rest, meaning that the source data has already been loaded into data storage. 

> Batch processing is generally performed over large, flat datasets that need to be prepared for further analysis. Log processing and data warehousing are common batch processing scenarios. In this scenario, you analyze information about GitHub projects, such as the number of time different projects have been forked or how recently projects have ben updated. 

Reviewer: ```suggestion
Batch processing is generally performed over large, flat datasets that need to be prepared for further analysis. Log processing and data warehousing are common batch processing scenarios. In this scenario, you analyze information about GitHub projects, such as the number of time different projects have been forked or how recently projects have been updated. 
```

In this tutorial, you learn how to:

> [!div class="checklist"]
>
> * Create and run a .NET for Apache Spark application
> * Read data into a DataFrame and prepare it for analysis
> * Process the data using Spark SQL

## Prerequisites 

If this is your first time using .NET for Apache Spark, check out the [Get started with .NET for Apache Spark](../tutorials/get-started.md) tutorial to learn how to prepare your environment and run your first .NET for Apache Spark application.

## Download the sample data

> [GHTorrent](http://ghtorrent.org/) monitors all public GitHub events, such as info about projects, commits, and watchers, and stores the events and their structure in databases. Data collected over different time periods is available as downloadable archives. Because the dump files are very large, this guide uses a [truncated version of the dump file](https://github.com/dotnet/spark/tree/master/examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/projects_smaller.csv) that can be downloaded from GitHub.

Reviewer: The PR that uploads this truncated file hasn't been merged yet, so this link wouldn't work today (but it will work once that PR is merged and the data is merged into the main .NET for Spark repo). So we'd just want to make sure this doc gets published when/after that PR is merged.

Reviewer: Also, per [GHTorrent guidelines](http://ghtorrent.org/faq.html), we might need to include licensing information like the following:

`The GHTorrent dataset is distributed under a dual licensing scheme ([Creative Commons +](https://wiki.creativecommons.org/wiki/CCPlus)). For non-commercial uses (including, but not limited to, educational, research or personal uses), the dataset is distributed under the [CC-BY-SA license](https://creativecommons.org/licenses/by-sa/4.0/).`

## Create a console application

1. In your command prompt, run the following commands to create a new console application:

   ```console
   dotnet new console -o mySparkBatchApp
   cd mySparkBatchApp
   ```

   The `dotnet` command creates a `new` application of type `console` for you. The `-o` parameter creates a directory named *mySparkBatchApp* where your app is stored and populates it with the required files. The `cd mySparkBatchApp` command changes the directory to the app directory you just created.

1. To use .NET for Apache Spark in an app, install the Microsoft.Spark package. In your console, run the following command:

>    ```console

Reviewer 2: We have an identifier for .NET CLI as well

```suggestion
   ```dotnetcli
```

   dotnet add package Microsoft.Spark
   ```

## Create a SparkSession

1. Add the following additional `using` statements to the top of the *Program.cs* file in *mySparkBatchApp*:

   ```csharp
   using System;
   using Microsoft.Spark.Sql;
   using static Microsoft.Spark.Sql.Functions;
   ```

> 1. Add the following code to establish a new SparkSession. The SparkSession is the entry point to programming Spark with the Dataset and DataFrame API. By calling the `spark` object, you can access Spark and DataFrame functionality throughout your program.

Reviewer 2: Add the following code where?

   ```csharp
   SparkSession spark = SparkSession
        .Builder()
        .AppName("GitHub and Spark Batch")
        .GetOrCreate();
   ```

## Prepare the data

> 1. Read the input file into a DataFrame, which is a distributed collection of data organized into named columns. You can set the columns for your data through <xref:Microsoft.Spark.Sql.DataFrame.Schema%2A>. Use <xref:Microsoft.Spark.Sql.DataFrame.Show%2A> to display the data in your DataFrame. Be sure to update the CSV filepath to the location of the GitHub data you downloaded.

Reviewer 2: ```suggestion
1. Read the input file into a `DataFrame`, which is a distributed collection of data organized into named columns. You can set the columns for your data through <xref:Microsoft.Spark.Sql.DataFrame.Schema%2A>. Use the <xref:Microsoft.Spark.Sql.DataFrame.Show%2A> method to display the data in your DataFrame. Be sure to update the CSV file path to the location of the GitHub data you downloaded.
```

   ```csharp
   DataFrame projectsDf = spark
       .Read()
       .Schema("id INT, url STRING, owner_id INT, " +
       "name STRING, descriptor STRING, language STRING, " +
       "created_at STRING, forked_from INT, deleted STRING" +
       "updated_at STRING")
       .Csv("filepath");

   projectsDf.Show();
   ```

> 1. Use <xref:Microsoft.Spark.Sql.DataFrame.Na%2A> to drop rows with NA (null) values, and the <xref:Microsoft.Spark.Sql.DataFrame.Drop%2A> method to remove certain columns from your data. This helps prevent errors if you try to analyze null data or columns that are not relevant to your final analysis.

Reviewer: Although the two uses of `xref` appear strange in the markdown, I assume they are normal-linking links/references when the doc gets published?

> 1. Use <xref:Microsoft.Spark.Sql.DataFrame.Na%2A> to drop rows with NA (null) values, and the <xref:Microsoft.Spark.Sql.DataFrame.Drop%2A> method to remove certain columns from your data. This helps prevent errors if you try to analyze null data or columns that are not relevant to your final analysis.

Reviewer 2: ```suggestion
1. Use the <xref:Microsoft.Spark.Sql.DataFrame.Na%2A> method to drop rows with NA (null) values, and the <xref:Microsoft.Spark.Sql.DataFrame.Drop%2A> method to remove certain columns from your data. This helps prevent errors if you try to analyze null data or columns that are not relevant to your final analysis.
```

   ```csharp
   // Drop any rows with NA values
   DataFrameNaFunctions dropEmptyProjects = projectsDf.Na();
   DataFrame cleanedProjects = dropEmptyProjects.Drop("any");

   // Remove unnecessary columns
   cleanedProjects = cleanedProjects.Drop("id", "url", "owner_id");
   cleanedProjects.Show();
   ```

## Analyze the data

1. Spark SQL allows you to make SQL calls on your data. It's common to combine user-defined functions and Spark SQL to apply a user-defined function to all rows of your DataFrame.

   You can specifically call `spark.Sql` to mimic standard SQL calls seen in other types of apps. You can also call methods like <xref:Microsoft.Spark.Sql.DataFrame.GroupBy%2A> and <xref:Microsoft.Spark.Sql.DataFrame.Agg%2A> to specifically combine, filter, and perform calculations on your data.

> 

Reviewer: Seems like it'd be worthwhile to explain what the code below is doing (since it'd be rather long or complex to a new user). Something like the following:

Our goal through this app is to gain some insights about our GitHub projects data. We start off by finding the number of times each language has been forked by grouping our data by language, and then calling the `Avg` method.

We next choose to sort our data in descending order through `Desc()` so that we can more easily see which languages are the most forked.

Next, we want to take a look at how recently the projects have been updated. We register a new user-defined function called _MyUDF_ and compare it with a date _s_referenceDate_ (that we had set earlier in our app) to see if each project was updated before or after that reference date. We use Spark SQL to then call that UDF on each row of our data and to thus analyze each project in our data set.

   ```csharp
   // Average number of times each language has been forked
   DataFrame groupedDF = cleanedProjects
       .GroupBy("language")

>        .Agg(Avg(cleanedProjects["forked_from"]) 

Reviewer 2: doesn't this need a ; at the end?

   // Sort by most forked languages first
   groupedDF.OrderBy(Desc("avg(forked_from)")).Show(); 

   spark.Udf().Register<string, bool>(
       "MyUDF",
       (date) => DateTime.TryParse(date, out DateTime convertedDate) &&
           (convertedDate > s_referenceDate);   
   cleanedProjects.CreateOrReplaceTempView("dateView"); 

   DataFrame dateDf = spark.Sql(
       "SELECT *, MyUDF(dateView.updated_at) AS datebefore FROM dateView");
   dateDf.Show();
   ```

1. Call `spark.Stop()` to end the SparkSession.

## Use spark-submit to run your app

1. Use the following command to build your .NET app:

   ```console
   dotnet build
   ```

> 1. Run your app with spark-submit. Be sure to update the following command with the actual paths to your Microsoft Spark jar file.

Reviewer 2: ```suggestion
1. Run your app with `spark-submit`. Be sure to update the following command with the actual paths to your Microsoft Spark jar file.
```

   ```console
   spark-submit --class org.apache.spark.deploy.dotnet.DotnetRunner --master local /<path>/to/microsoft-spark-<version>.jar dotnet /<path>/to/netcoreapp<version>/GitHubProjects.dll
   ```

## Get the code

You can see the [full solution](https://github.com/dotnet/spark/blob/master/examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/GitHubProjects.cs) on GitHub.

## Additional resources

> Provide links to where you can learn more about the topics covered in this tutorial

Reviewer: Want to remove this section? Or add something to it?

Author: I meant to remove it. Thanks!

## Next steps

> Advance to the next article to learn how to process streaming data with .NET for Apache Spark.

Reviewer: Thanks so much for your work on this! Marking as approved.
